students/jeff_shabani/session05/mailroom_3.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+import os
+import sys
+
+from collections import OrderedDict
+from operator import itemgetter
+from pathlib import Path
+
+DONORS = {'William B': [120, 130, 50],
+          'Sammy Maudlin': [500, 125, 670, 1000],
+          'Bobby Bittman': [10],
+          'Skip Bittman': [75, 125, 19],
+          'Ashley Lashbrooke': [10000, 15000]}
+
+prompt = "\n".join(("Welcome to my charity!",
+                    "Please select and option below:",
+                    "1 - Send a Thank You to an individual",
+                    "2 - Create a Report",
+                    "3 - Send letters to all donors",
+                    "4 - Quit",
+                    ">>> "))
+
+def value_error():
+    """
+    Catch non-numeric entries for donation amount.
+    """
+    amount = int()
+    while not amount:
+        try:
+            amount = int(input('How much would this donor like to donate?'))
+        except ValueError:
+            print('Please enter a valid numerical amount.')
+        else:
+            return amount
+
+
+def view_donor_names():
+    for name in DONORS:
+        print(name)
+
+
+def write_a_letter(name, amount):
+    letter = f'Dear {name},\n\nThank you for your kind donation of ${amount:,.0f}\n\n' \
+        f'Rest assured that these funds will be put to optimal use.\n\n' \
+        f'Best regards,\n' \
+        f'The Charitable Charities Team'
+    return letter
+
+
+def create_directory_decision():
+    """
+    Ask user if they want the directory they entered
+    created if it does not currently exsit.
+    """
+    answer = input('Path does\'t exist. Do you want to create it?')
+    return answer
+
+
+def validate_letter_directory_path():
+    """
+    Check if user-entered directory exists and offer them the
+    choice to create it if not. Default is current working
+    directory
+    """
+    location = input(f'Please enter the full path of the directory\n'
+                     f'where you want to save your letters.\n'
+                     f'Hit <Enter> to save to the current working directory.')
+
+    if not location:
+        location = os.getcwd()
+    else:
+        location = Path(f'{location}')
+        if not location.exists():
+            create_directory_answer = create_directory_decision()
+            #accept any version of yes, yep, etc.
+            if create_directory_answer.lower().startswith('y'):
+                location.mkdir()
+                location = os.chdir(location)
+            else:
+                location = os.getcwd()
+        else:
+            location = os.chdir(location)
+    return location
+
+
+def add_donations_and_send_thank_you():
+    while True:
+
+        answer = input('Please enter a donor Full Name.')
+
+        if answer.lower() == 'list':
+            view_donor_names()
+            continue
+
+        amount = value_error()
```

Comment:
the naming of this function is odd. its as if you want a value error exception to be held in the amount variable

Reply:
Probably better to name it as `get_amount()`

Reply:
Actually I would suggest a function that does this instead
```
def get_int_value(text):
    while True:
        try:
            value = int(input(text))
            return value
        except ValueError:
            print("Invalid value.  Please try again")
            continue
```

You can then call it as `amount = get_int_value("How much would this donor like to donate?")`

Even better might be to abstract it even more:
```
def get_value(text, check_type):
    value = None
    while True:
        try:
            value = check_type(input(text))
            return value
        except ValueError:
            print("Invalid value.  Please try again")
            continue
```
```
>>> get_value("Give me an integer: ", int)
Give me an integer: 4.5
Invalid value.  Please try again
Give me an integer: asdf
Invalid value.  Please try again
Give me an integer: 4
4
>>> get_value("Give me a float: ", float)
Give me a float: 3
3.0
>>> get_value("Give me a float: ", float)
Give me a float: asdf
Invalid value.  Please try again
Give me a float: !@
Invalid value.  Please try again
Give me a float: 10.0
10.0
```

Reply:
Notice that i'm not passing `int()` but `int`.  A reference to the function rather than calling the actual function until i'm in my loop.

Reply:
So check_type is a function created in the 'value = check_type(input(text))' . When 'get_value' is called in the 1st example,  the line becomes 'value = int(float)' which returns a value error because an int is not a float. Do I understand this correctly? I'm going to try running this through the debugger and see what happens.

Reply:
check_type is a reference to the function passed into get_value.  in the first example it doesn't become `value = int(float)`.  It becomes `value = input(int("Give me an number: "))`  If the user passes `3.5` or `asdf` the `ValueError` exception is thrown.

Reply:
I see. This is because value must be an integer, and throws and error in a non-integer is passed.